pisa/stages/absorption/pi_earth_absorption.py

```diff
@@ -133,7 +149,7 @@ def setup_function(self):
                                              'nuebar_nc', 'numubar_nc', 'nutaubar_nc'])
 
         for container in self.data:
-            self.layers.calcLayers(container['true_coszen'].get(WHERE))
+            self.layers.calcLayers(container['true_coszen'].get('host'))
```

Comment:
Is this necessary to run on gpu, to force this to be done on cpu? And is this a performance issue? I don't fully understand it but obviously this stands out

Reply:
The layers module fails on the GPU target apparently, and this is done the same way in the prob3 stage (but only once during setup).

Reply:
Fair enough. Don't you have to mark changed after an operation on the smart array, too?

Reply:
The layers module seems to be a host-only function. Since the containers for densities and distances are only created once, they don't have to be marked changed. However, it was still a good call, because the `rho_int` container does indeed change and needs to be marked as such.

Reply:
I'd still advocate to marke stuff as changed no matter if it is a special case like this where we can get away without doing it. It is just good practice and it does not create any overhead. The copy to and from device are lazy evaluations, so they only get executed when the array is needed in a subsequent calculation.